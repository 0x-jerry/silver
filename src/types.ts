/**
 *
 * `#flag` syntax is not support yet.
 *
 * @example -o --option @type:defaultValue #flag, description
 */
export interface CmdOption {
  name?: string
  alias?: string

  type?: string
  defaultValue?: string | boolean | number

  description?: string

  /**
   * Not support
   *
   * @todo implement it
   *
   * {@link OptionFlag}
   */
  flags?: string[]
}

/**
 * @example [@type:name:defaultValue]        // optional parameter
 * @example <@type:name:defaultValue>        // required parameter
 * @example [...@type:name]                  // array parameters
 */
export interface CmdParameter {
  name: string
  required?: boolean
  type?: string
  defaultValue?: string | boolean | number
  handleRestAll?: boolean
}

export interface Command {
  name: string
  alias?: string

  /**
   * action identifier
   */
  action?: string

  description?: string

  options?: CmdOption[]

  /**
   * {@link CommandFlag}
   */
  flags?: string[]

  parameters?: CmdParameter[]

  commands?: Command[]
}

export interface Program {
  /**
   * {@link ProgramFlag}
   */
  flags?: string[]

  command: Command

  actions?: Map<string, CmdAction>
}

export interface ActionParsedArgs {
  [arg: string]: any

  '--': string[]

  _: string[]
}

export interface CmdAction {
  <T extends ActionParsedArgs>(params: string[], opt: T): any
  (params: string[], opt: any): any
}

// --- enums

export enum CommandFlag {
  /**
   * stop early when parse argv
   */
  StopEarly = 'stopEarly',
}

export enum ProgramFlag {
  /**
   * Manual execute
   */
  Manual = 'manual',

  /**
   * Add `completion` sub command
   */
  Autocompletion = 'autocompletion',
}

export enum OptionFlag {}

export type CompletionValue = string | { label: string; desc?: string }

export type CompletionType = Array<CompletionValue>

export enum BuiltinType {
  File = '_files',
}
