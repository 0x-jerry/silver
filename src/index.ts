export { sliver } from './core'
export * from './types'
