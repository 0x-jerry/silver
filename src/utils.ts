import { toArray } from '@0x-jerry/utils'

export function splitFirst(text: string, separator: string) {
  const idx = text.indexOf(separator)

  return idx > -1 ? [text.slice(0, idx), text.slice(idx + separator.length)] : [text]
}

export const builtinType = {
  boolean: ['bool', 'boolean'],
  number: ['number'],
  string: ['string'],
}

export function isType(targetType: string | undefined, type: string | string[]) {
  return toArray(type).includes(targetType!)
}

export function isBuiltinType(type?: string) {
  return Object.values(builtinType).flat().includes(type!)
}
