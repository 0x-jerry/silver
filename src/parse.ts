import { createAutoIncrementGenerator, isPrimitive } from '@0x-jerry/utils'
import { CliConf, CliOption, CliProgram } from './types'

const TOKEN_ID_PREFIX = '$x@x$'
const nextId = createAutoIncrementGenerator(TOKEN_ID_PREFIX)

const tokenIdReg = new RegExp(`${TOKEN_ID_PREFIX}\\d+$`)

export function defineCli(raw: TemplateStringsArray, ...tokens: any[]): CliProgram {
  const tokenMapper = new Map<string, Function>()
  const tokenIdMapper = new Map<Function, string>()

  tokens.forEach((token) => {
    if (isPrimitive(token)) {
      return
    }

    const id = nextId()
    tokenMapper.set(id, token)
    tokenIdMapper.set(token, id)
  })

  const finalStr = raw.reduce((pre, cur, idx) => {
    const token = tokens[idx]

    const tokenId = tokenIdMapper.get(token)

    return pre + cur + (tokenId ? `${tokenId}` : token ?? '')
  }, '')

  let appCli: CliConf | null = null
  let currentCli: CliConf | null = null

  const lines = finalStr.trim().split('\n+')

  for (let line of lines) {
    line = line.trim()

    // skip blank string
    if (!line) continue

    if (line.startsWith('-')) {
      // is an option description

      const opt = parseOption(line)

      if (!currentCli) {
        throw new Error(`Invalid cli option, please specify a cli name first. ${line}`)
      }
      currentCli.options ||= []
      currentCli.options.push(opt)
      // todo, check duplicate name?
    } else {
      // is a cli description

      if (!appCli) {
        appCli = parseCli(line)
        currentCli = appCli
      } else {
        currentCli = parseCli(line)

        appCli.subCommands ||= []
        appCli.subCommands.push(currentCli)
      }
    }
  }

  if (!appCli) throw new Error('Parse CLI description failed!')

  const program = parseAction(appCli)

  return program

  function parseAction(cliConf: CliConf): CliProgram {
    const program: CliProgram = {
      ...cliConf,
      subCommands: undefined,
    }

    for (const subCliConf of cliConf.subCommands || []) {
      const subProgram = parseAction(subCliConf)

      program.subCommands ||= []
      program.subCommands.push(subProgram)
    }

    if (!cliConf.description) {
      return program
    }

    const [tokenId] = cliConf.description.trim().match(tokenIdReg) || []

    if (!tokenId) {
      return program
    }

    const fn = tokenMapper.get(tokenId)

    program.action = fn

    return program
  }
}

/**
 * Cli option description example:
 *
 * ```txt
 * -n --number #global @number:123.34, an number option with default value, and it's a global option.
 * ----------- ------- --------------  --------------
 * required    opt-in  opt-in          opt-in
 * ```
 */
function parseOption(description: string): CliOption {
  const conf: CliOption = {
    name: '',
  }

  const [opt, desc] = description.split(',')

  const segments = opt.split(/\s+/)

  segments.forEach((segment) => {
    if (segment.startsWith('-')) {
      // is name

      if (segment.startsWith('--')) {
        conf.name = segment.slice(2)
      } else {
        conf.alias = segment.slice(1)
      }
    } else if (segment.startsWith('#')) {
      // is flag

      conf.flags ||= []
      conf.flags.push(segment.slice(1))
    } else if (segment.startsWith('@')) {
      // is type

      const [type, defaultValue] = segment.split(':')

      conf.type = type.slice(1)

      if (['bool', 'boolean'].includes(conf.type!)) {
        conf.defaultValue = defaultValue === 'true'
      } else if (conf.type === 'number') {
        const n = parseFloat(defaultValue)
        conf.defaultValue = Number.isNaN(n) ? undefined : n
        // todo, warning ?
      } else {
        conf.defaultValue = defaultValue
      }
    } else {
      // unknown, just ignore
    }
  })

  conf.description = desc

  // fix name
  if (conf.alias && !conf.name) {
    conf.name = conf.alias
  }

  // check name
  if (!conf.name) {
    throw new Error(`Parse cli option failed: ${description}`)
  }

  return conf
}

/**
 * example:
 *
 * ```txt
 * commandName, A library for create command line interface quickly. ${command}
 * ```
 *
 */
function parseCli(description: string): CliConf {
  const conf: CliConf = {
    name: '',
  }

  const [name, desc] = description.split(',')

  conf.name = name

  conf.description = desc

  if (!name) {
    throw new Error(`Parse cli description failed: ${description}`)
  }

  return conf
}
