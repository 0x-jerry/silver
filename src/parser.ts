import { createAutoIncrementGenerator, isPrimitive } from '@0x-jerry/utils'
import type { Command, CmdOption, CmdParameter, Program, CmdAction } from './types'
import { builtinType, isType, splitFirst } from './utils'

const TOKEN_ID_PREFIX = '__token_id__'
const nextId = createAutoIncrementGenerator(TOKEN_ID_PREFIX)

const tokenIdReg = new RegExp(`${TOKEN_ID_PREFIX}\\d+`, 'g')

export function parseCliProgram(raw: TemplateStringsArray, ...tokens: any[]): Program {
  const actionMapper = new Map<string, CmdAction>()
  const actionIdMapper = new Map<Function, string>()

  tokens.forEach((token) => {
    if (isPrimitive(token)) {
      return
    }

    // skip if exists
    if (actionIdMapper.has(token)) {
      return
    }

    const id = nextId()
    actionMapper.set(id, token)
    actionIdMapper.set(token, id)
  })

  const finalStr = raw.reduce((pre, cur, idx) => {
    const token = tokens[idx]

    const tokenId = actionIdMapper.get(token)

    return pre + cur + (tokenId ? ` ${tokenId} ` : token ?? '')
  }, '')

  const conf = parseProgram(finalStr)

  conf.actions = actionMapper

  return conf
}

function parseProgram(finalStr: string): Program {
  const program: Partial<Program> = {}

  let mainCommand: Command | undefined
  let currentCommand: Command | undefined

  const lines = finalStr.trim().split(/\n+/)

  {
    // check program flags
    const flags = parseAppOption(lines[0])

    if (flags) {
      program.flags = flags
      lines.splice(0, 1)
    }
  }

  for (let line of lines) {
    line = line.trim()

    // skip blank string
    if (!line) continue
    // ignore comments
    if (line.startsWith('#')) continue

    if (line.startsWith('-')) {
      // is an option description
      const opt = parseOption(line)

      if (!currentCommand) {
        throw new Error(`Invalid cli option, please specify a cli name first. ${line}`)
      }

      currentCommand.options ||= []
      currentCommand.options.push(opt)
      // todo, check duplicate name?
    } else {
      // is a cli description
      if (!mainCommand) {
        mainCommand = parseCommand(line)
        currentCommand = mainCommand
      } else {
        currentCommand = parseCommand(line)

        mainCommand.commands ||= []
        mainCommand.commands.push(currentCommand)
      }
    }
  }

  program.command = mainCommand

  if (!program.command) {
    throw new Error('Parse CLI description failed!')
  }

  return program as Program
}

/**
 * Cli option description example:
 *
 * ```txt
 * -n --number #global @number:123.34, an number option with default value, and it's a global option.
 * ----------- ------- --------------  --------------
 * required    opt-in  opt-in          opt-in
 * ```
 */
function parseOption(description: string): CmdOption {
  const conf: CmdOption = {
    name: '',
  }

  const [opt, desc] = splitFirst(description, ',')

  const segments = opt.split(/\s+/)

  segments.forEach((segment) => {
    if (segment.startsWith('-')) {
      // is name

      if (segment.startsWith('--')) {
        conf.name = segment.slice(2)
      } else {
        conf.alias = segment.slice(1)
      }
    } else if (segment.startsWith('#')) {
      // is flag

      conf.flags ||= []
      conf.flags.push(segment.slice(1))
    } else if (segment.startsWith('@')) {
      // is type

      const [type, defaultValue] = splitFirst(segment, ':')

      conf.type = type.slice(1)

      if (isType(conf.type, builtinType.boolean)) {
        conf.defaultValue = defaultValue === 'true'
      } else if (isType(conf.type, builtinType.number)) {
        const n = parseFloat(defaultValue)
        conf.defaultValue = Number.isNaN(n) ? undefined : n
        // todo, warning ?
      } else {
        conf.defaultValue = defaultValue
      }
    } else {
      // unknown, just ignore
    }
  })

  conf.description = desc?.trim()

  // check name
  if (!conf.name && !conf.alias) {
    throw new Error(`Parse cli option failed: ${description}`)
  }

  return conf
}

/**
 * @example
 *
 * ```txt
 * aliasName/commandName <@file:dir> [...files] #flag, A library for create command line interface quickly. ${command}
 *
 * up/upgrade <@pkg:name:defaultPkgName> [version:latest] #flag, Upgrade package to the latest.
 * ```
 *
 */
function parseCommand(description: string): Command {
  const conf: Command = {
    name: '',
  }

  // extract action identifier, use the last one if it has multiple.
  description = description.replace(tokenIdReg, (id) => {
    conf.action = id
    return ''
  })

  const [name, desc = ''] = splitFirst(description, ',')

  const [nameWithAlias, ...parameters] = name.split(/\s+/g)

  // up/upgrade
  // upgrade
  {
    const names = splitFirst(nameWithAlias, '/')

    if (names.length === 2) {
      conf.alias = names[0]
      conf.name = names[1]
    } else {
      conf.name = nameWithAlias
    }
  }

  // parameters
  {
    parameters.forEach((parameter) => {
      if (parameter.startsWith('#')) {
        // is flag
        conf.flags ||= []
        conf.flags.push(parameter.slice(1))
        return
      }

      const p = parseCommandParameter(parameter)

      if (p) {
        conf.parameters ||= []
        conf.parameters.push(p)
      }
    })
  }

  conf.description = desc.trim()

  if (!name) {
    throw new Error(`Parse command description failed: ${description}`)
  }

  return conf
}

/**
 *
 * <@pkg:name:defaultPkgName>
 * [@type:version:latest]
 * [...@file:files]
 *
 * @param description
 */
function parseCommandParameter(description: string): CmdParameter | null {
  const l = description.slice(0, 1)
  const r = description.slice(-1)

  const conf: CmdParameter = {
    name: '',
  }

  if (l === '<' && r === '>') {
    conf.required = true
  } else if (l === '[' && r === ']') {
    conf.required = false
  } else {
    // invalid
    return null
  }

  let content = description.slice(1, -1)

  if (content.startsWith('...')) {
    conf.isArray = true
    content = content.slice(3)
  }

  const segments = content.split(':')

  const hasType = segments.at(0)?.startsWith('@')

  if (hasType) {
    conf.type = segments[0].slice(1)
    segments.splice(0, 1)
  }

  const [name, defaultValue] = segments

  conf.name = name
  conf.defaultValue = defaultValue

  return conf
}

/**
 *
 * @param description
 * @example
 *
 * ```txt
 * @manual @help
 * ```
 */
function parseAppOption(description: string) {
  if (!description.trim().startsWith('@')) return null

  const flags: string[] = []

  const segments = description.trim().split(/\s+/g)

  segments.forEach((item) => {
    if (!item.startsWith('@')) {
      return
    }

    flags.push(item.slice(1))
  })

  return flags
}
