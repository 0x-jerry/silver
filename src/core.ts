import { type Value, sleep, textTableToString, isString, toValue } from '@0x-jerry/utils'
import { parseCliProgram } from './parser'
import {
  type ActionParsedArgs,
  type Program,
  CommandFlag,
  type Command,
  ProgramFlag,
  type CompletionType,
} from './types'
import minimist from 'minimist'
import { isType, builtinType } from './utils'
import { generateZshAutoCompletion, normalizeStr } from './completion/zsh'

export class Sliver {
  conf?: Program

  typeMapper = new Map<string, Value<CompletionType | Promise<CompletionType>>>()

  parse(raw: TemplateStringsArray, ...tokens: any[]) {
    this.conf = parseCliProgram(raw, ...tokens)

    if (this.conf.flags?.includes(ProgramFlag.Help)) {
      // add help option
      this.conf.command.options ||= []

      this.conf.command.options.push({
        name: 'help',
        alias: 'h',
        type: 'bool',
        description: 'Print help messages.',
      })
    }

    if (this.conf.flags?.includes(ProgramFlag.Autocompletion)) {
      const { cmd, action } = createCompletionCommand(this)

      this.conf.command.commands ||= []
      this.conf.command.commands.push(cmd)

      this.conf.actions ||= new Map()

      this.conf.actions.set(cmd.action!, action)
    }
  }

  async getCompletion(type: string) {
    const value = this.typeMapper.get(type)

    return (await toValue(value)) || []
  }

  type(name: string, getType: Value<CompletionType | Promise<CompletionType>>) {
    this.typeMapper.set(name, getType)

    return this
  }

  execute(argv: string[]) {
    if (!this.conf) return

    // get all sub command names
    const commandMapper = this.conf.command.commands?.reduce((names, item) => {
      names.set(item.name, item)

      if (item.alias) {
        names.set(item.alias, item)
      }

      return names
    }, new Map<string, Command>())

    let command = this.conf.command

    // if it is a sub command
    if (commandMapper?.has(argv[0])) {
      command = commandMapper.get(argv[0])!
      argv = argv.slice(1)
    }

    const args = parseArgv(argv, command)

    if (this.conf.flags?.includes(ProgramFlag.Help) && (args.help || args.h)) {
      // force to print help message
      console.log(generateHelpMsg(command))

      return
    }

    // todo, validate required parameters and options

    const action = this.conf.actions?.get(command.action!)
    action?.(args._, args)
  }
}

/**
 *
 * @example
 *
 * ```ts
 * const ins = sliver`@help @autocompletion
 *
 * xcat [...dir], Read file like \`cat\` command.
 *
 * -o --option \@bool, Global option.
 *
 * run [@scripts:scriptName], Run some script.
 *
 * -p --pattern \@string, Use glob pattern to run multiple scripts.
 * `
 *
 * ins.type('scripts', ['script1', 'script2']) // register auto completion for shell
 * ```
 *
 */
export function sliver(raw: TemplateStringsArray, ...tokens: any[]) {
  const ins = new Sliver()

  ins.parse(raw, ...tokens)

  if (!ins.conf?.flags?.includes(ProgramFlag.Manual)) {
    sleep().then(() => {
      const argv = process.argv.slice(2)
      ins.execute(argv)
    })
  }

  return ins
}

function parseArgv(argv: string[], program: Command) {
  const config = {
    alias: {} as Record<string, string>,
    default: {} as Record<string, any>,
    boolean: [] as string[],
    '--': true,
    stopEarly: !!program.flags?.includes(CommandFlag.StopEarly),
  }

  for (const opt of program.options || []) {
    if (opt.alias && opt.name) {
      config.alias[opt.name] = opt.alias
    }

    if (opt.defaultValue != null) {
      const nameOrAlias = opt.name || opt.alias
      if (nameOrAlias) {
        config.default[nameOrAlias] = opt.defaultValue
      }
    }

    if (isType(opt.type!, builtinType.boolean)) {
      const nameOrAlias = opt.name || opt.alias
      if (nameOrAlias) {
        config.boolean.push(nameOrAlias)
      }
    }
  }

  const args = minimist(argv, config)

  args['--'] ||= []

  return args as ActionParsedArgs
}

function generateHelpMsg(conf: Command) {
  const msgs: string[] = []

  const usageDescription = conf.commands?.length ? `[COMMAND] [OPTIONS]` : `[OPTIONS]`

  const composeAlias = (cmd: Command) => [cmd.alias, cmd.name].filter(Boolean).join('/')
  const usage = `${composeAlias(conf)} ${usageDescription} ${conf.description}`

  msgs.push(usage, '')

  if (conf.commands?.length) {
    const commands = conf.commands.map((item) => [composeAlias(item), item.description])

    const s = textTableToString(commands)
    msgs.push(s, '')
  }

  if (conf.options?.length) {
    const options = conf.options.map((item) => {
      const name =
        item.alias && item.alias !== item.name ? `--${item.name} -${item.alias}` : `--${item.name}`
      return [name, item.description]
    })

    const s = textTableToString(options)
    msgs.push(s, '')
  }

  return msgs.join('\n')
}

function createCompletionCommand(ins: Sliver) {
  const conf = parseCliProgram`
completion [type], Generate autocompletion for zsh.

--install, Install autocompletion for zsh, not implement yet.
--uninstall, Uninstall autocompletion for zsh, not implement yet.
  `

  conf.command.action = 'completion'

  return {
    cmd: conf.command,
    action,
  }

  async function action(params: string[], opt: { install?: boolean; uninstall?: boolean }) {
    const [type] = params

    if (type) {
      const completions = await ins.getCompletion(type)

      const s = completions
        .map((item) => {
          return isString(item)
            ? normalizeStr(item)
            : item.desc
            ? `${normalizeStr(item.label)}\\:${item.desc}`
            : normalizeStr(item.label)
        })
        .join('\n')
      process.stdout.write(s)
      return
    }

    if (opt.install) {
      return
    }

    if (opt.uninstall) {
      return
    }

    if (ins.conf?.command) {
      const zshCode = generateZshAutoCompletion(ins.conf?.command)

      console.log(zshCode)
    }
  }
}
