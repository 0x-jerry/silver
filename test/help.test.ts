import type { SpyInstance } from 'vitest'
import { sliver } from '../src'

describe('test help message', () => {
  const ins = sliver`
@help @autocompletion @manual

xx [@type:type], A library for create command line interface quickly.

-t --test @test:t1, Test autocompletion.

# aliasName/commandName
up/upgrade <@test:dir> [...other] #stopEarly, an sub command.

-s --string @string:default, sub command option.
-sm --small @bool, other option.
`

  const ctx = {
    logs: [] as any[],
    log: null as null | SpyInstance<any[], void>,
  }

  beforeEach(() => {
    ctx.logs = []

    ctx.log = vi.spyOn(console, 'log').mockImplementation((...msg: any) => {
      ctx.logs.push(...msg)
    })
  })
  afterEach(() => {
    ctx.log?.mockClear()
    ctx.logs = []
  })

  it('should print main help message', () => {
    ins.execute(['-h'])

    expect(ctx.logs).toMatchSnapshot()
  })

  it('should print sub command help message', () => {
    ins.execute(['up', '-h'])
    expect(ctx.logs).toMatchSnapshot()
  })
})
