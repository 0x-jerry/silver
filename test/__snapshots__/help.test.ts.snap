// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`test help message > should print main help message 1`] = `
[
  "xx [COMMAND] [OPTIONS] A library for create command line interface quickly.

up/upgrade an sub command.                 
completion Generate autocompletion for zsh.

--test -t Test autocompletion.
--help -h Print help messages.
",
]
`;

exports[`test help message > should print sub command help message 1`] = `
[
  "up/upgrade [OPTIONS] an sub command.

--string -s sub command option.
--small -sm other option.      
",
]
`;
